python/tests/test_topology.py

```diff
@@ -904,6 +903,398 @@ def test_internal_samples(self):
+            span = t1_diffs[0][1] - left
+            total += norm_kc_vectors(t1_vecs, t2_vecs, lambda_) * span
+
+            left = t1_diffs[0][1]
+            t1, t1_diffs = next(tree_iter_1), next(edge_diffs_iter_1)
+            update_kc_incremental(t1, t1_vecs, t1_diffs, sample_maps[0])
+        span = t2_diffs[0][1] - left
+        left = t2_diffs[0][1]
+        total += norm_kc_vectors(t1_vecs, t2_vecs, lambda_) * span
+
+    return total / ts1.sequence_length
+
+
+def update_kc_incremental(tree, kc, edge_diffs, sample_index_map):
+    _, edges_out, edges_in = edge_diffs
+
+    # Update any depths/times changed by removed edges.
+    for e in edges_out:
+        visit_in_subtree_pairs(tree, kc, e.child, sample_index_map)
+    for e in edges_in:
+        if tree.is_leaf(e.child):
+            u = e.child
+            time = tree.branch_length(u)
+            update_kc_vectors_single_leaf(kc, u, 1, time, sample_index_map)
+        visit_in_out_subtree_pairs(tree, kc, e, sample_index_map)
+        visit_in_subtree_pairs(tree, kc, e.child, sample_index_map)
+
+
+def visit_in_subtree_pairs(tree, kc, u, sample_index_map):
+    """
+    Visit all pairs of leaves of the internal node u, tracking
+    the mrca of each pair and the depth of that mrca.
+    """
+    root_time = tree.time(tree.root)
+    stack = [(u, node_depth(tree, u))]
+    while len(stack) > 0:
+        u, depth = stack.pop()
+        if not tree.is_leaf(u):
+            c1 = tree.left_child(u)
+            while c1 != -1:
+                stack.append((c1, depth + 1))
+                c2 = tree.right_sib(c1)
+                while c2 != -1:
+                    for l1 in tree.leaves(c1):
+                        for l2 in tree.leaves(c2):
+                            time = root_time - tree.time(u)
+                            update_kc_vectors_pair(
+                                kc, l1, l2, depth, time, sample_index_map
+                            )
+                    c2 = tree.right_sib(c2)
+                c1 = tree.right_sib(c1)
+
+
+def visit_in_out_subtree_pairs(tree, kc, edge, sample_index_map):
+    """
+    Visit all pairs of leaves with one leaf in the subtree under
+    the given edge and the other leaf outside the subtree. Track
+    the mrca of each pair of leaves and the depth/time-to-root of
+    that mrca.
+    """
+    root_time = tree.time(tree.root)
+    edge_c = edge.child
+    p = edge.parent
+    depth = node_depth(tree, p)
+    c = edge_c
+    while p != -1:
+        for sibling in tree.children(p):
+            if sibling != c:
+                for out_leaf in tree.leaves(sibling):
+                    for in_leaf in tree.leaves(edge_c):
+                        time = root_time - tree.time(p)
+                        update_kc_vectors_pair(
+                            kc, out_leaf, in_leaf, depth, time, sample_index_map
+                        )
+        c, p = p, tree.parent(p)
+        depth -= 1
+
+
+def node_depth(tree, u):
```

Comment:
Would it be useful elsewhere if we added this to the `Tree` class?

Reply:
Yes! ``Tree.depth(u)`` which returns the length of the path from u to root would be a useful operation I think. Would you mind following it up?